examples/run/cuda/bfield_cuda.cu

```diff
@@ -0,0 +1,625 @@
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+        0,
+    };
+
+    /*
+     * Bitmap rows must be padded to multiples of 4, so we will keep these
+     * padding zeros handy for when we need to do that.
+     */
+    char padding[3] = {0, 0, 0};
+
+    /*
+     * Start off by writing the 54 byte header to the file.
+     */
+    bmp.write(header, 54);
+
+    /*
+     * Here we compute the colour palette. Each byte identifies one of 256
+     * colours, and this bit of code defines what those colors are. We use a
+     * part of the HSV spectrum which looks nice. For an explanation of this
+     * code look up the code for HSV to RGB conversion.
+     */
+    for (unsigned int i = 0; i < 256; ++i) {
+        float s = 0.7;
+        float v = 1.0;
+        float hue = (i / 256.0) * 240.0;
+        float c = v * s;
+        float hp = hue / 60.0;
+
+        float x = c * (1 - std::abs(std::fmod(hp, 2.0) - 1.0));
+
+        float rp, gp, bp;
+
+        if (hp < 1.0) {
+            rp = c, gp = x, bp = 0;
+        } else if (hp < 2.0) {
+            rp = x, gp = c, bp = 0;
+        } else if (hp < 3.0) {
+            rp = 0, gp = c, bp = x;
+        } else if (hp < 4.0) {
+            rp = 0, gp = x, bp = c;
+        } else if (hp < 5.0) {
+            rp = x, gp = 0, bp = c;
+        } else if (hp < 6.0) {
+            rp = c, gp = 0, bp = x;
+        }
+
+        float q = v - c;
+
+        /*
+         * For each of the 256 colours, write the output RGB colour to the file
+         * in four bytes.
+         */
+        char cmap[4] = {
+            static_cast<char>((rp + q) * 255),
+            static_cast<char>((gp + q) * 255),
+            static_cast<char>((bp + q) * 255),
+            0,
+        };
+
+        bmp.write(cmap, 4);
+    }
+
+    /*
+     * Finally, write the image section. Start off at the bottom of the image
+     * and work our way up.
+     */
+    for (std::size_t y = h - 1; y < h; --y) {
+        for (std::size_t x = 0; x < w; ++x) {
+            /*
+             * Retrieve the magnitude from the image, and write it.
+             */
+            char r = img[h * x + y];
+            bmp.write(&r, 1);
+        }
+
+        /*
+         * If necessary, write the appropriate padding to the file.
+         */
+        bmp.write(padding, (4 - (w % 4)) % 4);
+    }
+
+    bmp.close();
+}
+
+/*
+ * This is the meat of the pudding, but it is very simple. Given a magnetic
+ * field and some coordinates, fetch the magnetic fiend at that point, possibly
+ * doing trilinear interpolation.
+ *
+ * Returns a 4-vector, but only the first 3 values are used.
+ */
+__device__ float4 read_bfield(MagneticField bf, float x, float y, float z) {
+    /*
+     * We take the coordinates, offset them as appropriate, and add 0.5 which
+     * is required by the interpolation hardware.
+     */
+    return tex3D<float4>(bf.texture, x + bf.offset_x + 0.5,
```

Comment:
If I understand correctly, you added 0.5 for triggering interpolation? (0.5 possibly half of b field cell interval?)
If so, I wonder what the interpolation algorithm is. 
Actually I want to have the same interpolation algorithm used by the simulation stage.  
If the interpolation algorithm is different, we may get the unexpected result after Kalman filtering

Reply:
Okay good. the linear filtering may be what we also have in Acts (someone needs to confirm though)